Tutorial: Let's talk to Kafka. How to send and receive application data from Apache Kafka®?

dewan-ahmed opened the pull request:
# What changed, and why it matters

This PR brings the files for **Let's talk to Kafka. How to send and receive application data from Apache Kafka®?** tutorial. 

For now, the PR is marked as DRAFT and I'd like to discuss the following:

- Can I add/improve the narrative?
- Can I update the Mermaid diagram based on the narrative?
- I didn't include talking to Kafka using the built-in CLI (kafka-console-producer.sh and kafka-console-consumer.sh). Do you think I could add that part before using Java and Python?

Please suggest any other improvements you find for this PR.

mattstratton commented:
For anyone looking for the direct preview link - https://tutorial-kafka-basics.devportal.pages.dev/docs/tutorials/talk-to-kafka

dewan-ahmed commented:
@TibsAtWork I recall we had a discussion earlier about using anchor tags with Sphinx. I had to modify the following links as the link check was causing the build to fail.

```
Under the hood, the libraries make use of the `Producer API <https://kafka.apache.org/documentation/#producerapi>`_ and the `Consumer API <https://kafka.apache.org/documentation/#consumerapi>`_. 
```

Is there a way to include links like https://kafka.apache.org/documentation/#producerapi on devportal?

TibsAtWork commented:
It's not ideal, but the fallback we have (which we've used for other kafka links) is to tell sphinx not to check them.
This is done in `conf.py` using `linkcheck_ignore` - see https://github.com/aiven/devportal/blob/tutorial-kafka-basics/conf.py#L82 and following on your branch.

mattstratton commented:
I think that it's still too much focused on Aiven setup, etc. I would make a few tweaks:

where it says "To get started, you’ll need:

An [Aiven account](https://console.aiven.io/signup)" this could be 

"you'll need an aiven account or your own kafka setup" etc

additionally, maybe riht before the "create an aiven for kafka service" something like "if you already have a kafka service, go ahead and jump to "create a topic" (with a link) (or whatever the right place to jump to is)

I think the flow needs to not assume everyone is using aiven to complete this tutorial :)

mattstratton commented:
also maybe rephrase `In this tutorial, we will learn how to create an Aiven for Apache Kafka® service`

mattstratton commented:
as we just discussed, maybe remove the "variables" section and just link to the docs as "if you want to know all the options" or something

mattstratton commented:
one other note - add the specific files for each "program" and then say "when you run `python consumer.py` you will see x in your console or whatever)

dewan-ahmed commented:
Thanks for the helpful feedback @mattstratton 

https://github.com/aiven/devportal/pull/1733/commits/14be81d78e2b7cb105ccf8bd6007beab1badb777 brings in the following changes based on your suggestion:
- rewrite the tutorial so that the learner can use a local Kafka instance or Aiven managed Kafka
- remove Aiven token as a pre-req. they just need kafka service - whether local or managed
- add detailed instructions on how to run the Python files and what would happen when they do

Next commits will bring in the section on Karapace. Once I add all the sections, I'll run the code samples and do a final proof-read.

dewan-ahmed commented:
Hi @ftisiot looking for your technical review on the Karapace section. I added it based on your review feedback. Feel free to review the whole tutorial if you have time 🙏🏽

ftisiot reviewed (changes requested):
Hi Dewan! This is very solid, some minor points you might want to address (I believe I fixed some of them already)
- I would spend more time describing the consumer group with a mermaid example. as of now it's all very "condensed"
- the "Setting up access control lists (ACLs) for Karapace schema registry and REST proxy" needs a series of (numbered) steps since there are multiple actions involved
- The Karapace bit it's interesting (probably it needs its own tutorial) but feels like it's "detached" from the main learning component. I would suggest to store the content for another tutorial and replace it with pushing and consuming avro messages from python (for this you might need to use the confluent libraries)

dewan-ahmed commented:
Hi @ftisiot thanks so much for the review. I'll go over these points next week and commit accordingly.

dewan-ahmed commented:
Thanks again for the review @ftisiot 

In the last few commits, I have:

1. Added more description for consumer group with a Mermaid diagram.
2. Removed Karapace piece from this blog and will create a separate article on that.
3. Added Avro piece. This took significantly long to get working for the code samples. AvroProducer and AvroConsumer are deprecated so I used AvroSerializer and AvroDeserializer. I still have an error with the Avro producer program and I'm working on to resolve that. Meanwhile, if you have some time, please run the Avro programs and test it in case you see the issue OR please point me to a consumer program for reference. 

```
Traceback (most recent call last):
  File "/Users/dewan.ahmed/code/scratch-work/kafka-tutorial/kafka-producer01/avro-producer-ssl.py", line 36, in <module>
    avro_serializer = AvroSerializer(schema_str, schema_registry_client)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/confluent_kafka/schema_registry/avro.py", line 203, in __init__
    schema = _schema_loads(schema_str)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/confluent_kafka/schema_registry/avro.py", line 61, in _schema_loads
    schema_str = schema_str.strip()
                 ^^^^^^^^^^^^^^^^
AttributeError: 'SchemaRegistryClient' object has no attribute 'strip'
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,579 @@
+Observation
+~~~~~~~~~~~
+
+In the above examples, we're using ``confluent-kafka-python`` library to send and consume Avro messages to/from a Kafka broker. You can follow the SSL protocol example to create producer and consumer codes for SASL protocol. Here's an overview of what each program does:
```
Why are you mentioning SASL here? should we give a hint on what they need to change for sASL?

dewan-ahmed replied:
Good catch. Removed mention and added suggestion.

ftisiot reviewed (changes requested):
Hi Dewan! Really good improvement.
I would switch Producer and consumer to logically first write and then read and change little things in the format.

But overall, really interesting piece!

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,579 @@
+
+With SSL authentication:
+
+.. code:: python
```
Made the change for producer/consumer swap. I'll be running the entire tutorial on a fresh VM now. then I can check if we need the learner to install any Python libraries.

dewan-ahmed replied:
Just verified; you're right. Added `fastavro` requirement via https://github.com/aiven/devportal/pull/1733/commits/c039a69367d93a036e2aedf6b29a4e6f9a4c51d0.

dewan-ahmed commented:
Please take another look @ftisiot 

I addressed the recent comments.

If this looks Ok, we can move to editorial review. Thanks.

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+Overview
+--------
+
+Getting started with Apache Kafka can be simple as you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
```
I would avoid `simple`
```suggestion
Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+--------
+
+Getting started with Apache Kafka can be simple as you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
+If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_.
```
```suggestion
If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_, allowing you to start producing and consuming data in minutes.
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+    pip install kafka-python
+
+- `confluent-Kafka <https://github.com/confluentinc/confluent-kafka-python>`_ library installed, you can install it with the following command:  
```
Why are we using both kafka python and confluent kafka?

```suggestion
- `confluent-Kafka <https://github.com/confluentinc/confluent-kafka-python>`_ library installed that you'll use to produce messages in AVRO format, you can install it with the following command:  
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+6. Click **Create Service** under the summary on the right side of the console.
+
+The blinking blue icon besides your service name will indicate that VMs are being provisioned. Once the icon turns solid green, your Apache Kafka service is up and running.
```
```suggestion
The blinking blue icon besides your service name will indicate that service is being provisioned. Once the icon turns solid green, your Apache Kafka service is up and running.
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+    bin/kafka-topics.sh --create --topic demo-topic --bootstrap-server localhost:9092
+
+For an Aiven for Apache Kafka service, you can create the topic from the Aiven console. 
+From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name "demo-topic". Click **Add topic**.
```
```suggestion
From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name ``demo-topic``. Click **Add topic**.
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+For an Aiven for Apache Kafka service, you can create the topic from the Aiven console. 
+From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name "demo-topic". Click **Add topic**.
+Once this topic is created, we can see that the number of partitions is 1. 
```
```suggestion
Once this topic is created, we can see that the default number of partitions is 1. 
```

ftisiot replied:
where do we see the # of partitions?

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+The concept of consumer group and consuming messages on Kafka
+------------------------------------------------------------------
+
+Consumer group is the logical grouping of consumers. In Kafka, the consumer(s) must belong to a consumer group, even if it's the default consumer group. 
```
```suggestion
Consumer group is the logical grouping of consumers. In Kafka, the consumer(s) must belong to a consumer group; if you don't define one, a random consumer group will be generated. 
```

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+    graph TD
+
+        A(Topic) -- Partition 3 --> D[/Consumer 3/]
+        A(Topic) -- Partition 4 --> E[/Consumer 4/]
+        subgraph Consumer Group 2
+        D
+        E
+        end
+
+        A(Topic) -- Partition 1 --> B[/Consumer 1/]
+        A(Topic) -- Partition 2 --> C[/Consumer 2/]
+        subgraph Consumer Group 1
+        B
+        C
+        end
```
I believe this is wrong, since it seems that is splitting the topic across consumers belonging to two different consumer groups

dewan-ahmed replied:
Thanks for the feedback. Fix is via https://github.com/aiven/devportal/pull/1733/commits/4c4cb085be4be48c0e95b7360ff8deec720d9a44.

ftisiot reviewed (changes requested):
We're almost there, I believe one of the main graphs is not correct tho :D

dewan-ahmed commented:
Hi @ftisiot please do one more review. I fixed the consumer group diagram based on your last review. Since the first tutorial was merged, I got some merge conflict on this PR and had to add the ``anomaly-detection.rst`` file. Please check the preview URL that both tutorials look Ok. Happy to address anything else that might need a fix.

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+4. Select the cloud provider and region that you want to run your service on.
+
+5. Select `business-4` as service plan.
```
Why do we need a `business-4`? If there's a particular requirement (like being able to enable Karapace) we should mention, otherwise we should select a minor plan

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+.. code:: bash
+
+    bin/kafka-topics.sh --create --topic demo-topic --bootstrap-server localhost:9092
```
We should mention that `localhost:9092` needs to be replaced with the server name

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+    bin/kafka-topics.sh --create --topic demo-topic --bootstrap-server localhost:9092
+
+For an Aiven for Apache Kafka service, you can create the topic from the Aiven console. 
+From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name ``demo-topic``. Click **Add topic**.
```
Can we add a screenshot here?

ftisiot replied:
And also can we have this as list of steps in different dots?

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+
+For an Aiven for Apache Kafka service, you can create the topic from the Aiven console. 
+From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name ``demo-topic``. Click **Add topic**.
+Once this topic is created, we can see that the default number of partitions is 1. 
```
Screenshot?

ftisiot commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,688 @@
+The producer program uses the ``confluent-kafka`` library to create an AvroProducer instance and send Avro messages to a Kafka broker. Here's a breakdown of what the program does:
+
+- Import the necessary libraries: ``confluent_kafka``, ``confluent_kafka.avro``, ``ssl``, and ``io``.
+- Define the URL of the Kafka broker and the Avro schema for the message.
+- Set up SSL configuration by creating an SSL context and loading the client certificate, key, and CA certificate.
+- Define the AvroProducer configuration, including the SSL settings and the Avro schema.
+- Create an AvroProducer instance.
+- Send a sample Avro message to the Kafka broker.
```
I believe we need to show the parts of the code which are in charge of this. The code is long and we need to link back to what we're doing.

Also we need to add the fact that we define the schema

dewan-ahmed replied:
Thanks @ftisiot 

I've addressed the other comments. A question regarding adding text regarding Avro piece by highlighting the code blocks. Will it be helpful if I add comments within the code blocks OR do you suggest I add the text before the actual code block?

ftisiot replied:
That is a very good suggestion!

dewan-ahmed replied:
https://github.com/aiven/devportal/pull/1733/commits/7abb8dab7dd9f9da81b4de816f68404c0a2e3b9d brings in comments within Avro code blocks to explain what each of the segments do. Please check if this looks Ok 🙏🏽

ftisiot reviewed (changes requested):
Hi Dewan!
Good progress, I believe we need to tell a bit more in detail the story about the Avro piece by highlighting the code blocks which are responsible of each section

We're almost there

ftisiot commented:
LGTM! One minor suggestion, remove `Tutorial:` from the main title. 
I believe the next phase before publishing is the editorial review!
Well done Dewan!

celestehorgan commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+Any format would do as long as it's consistent. For this tutorial, we're selecting Apache Avro which is an open-source project and one of the most popular serialization format.
+Avro is defined by a schema and the schema is written in JSON. You can consider Avro as JSON with a schema attached to it.
+
+Both of these programs use SSL authentication. You can follow the previous examples of this tutorial if you'd like to use SASL authentication instead.
```
```suggestion
Both of these programs use SSL authentication. You can follow the previous examples in this tutorial if you'd like to use SASL authentication instead.
```

celestehorgan commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here's an example of a Kafka consumer in Python using the `confluent-Kafka-python <https://github.com/confluentinc/confluent-kafka-python>`_ library to produce Avro-encoded messages.
+On a terminal window, run the following producer code. The program should execute and exit. You won't see anything yet since we need the consumer program to run next. 
```
```suggestion
In a terminal window, run the following producer code. The program should execute and exit. You won't see anything yet since we need the consumer program to run next. 
```

celestehorgan commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+
+Here's an overview of what each program does:
+
+Producer program
```
Put the content in this section (as in: copy and paste it) in L#380 where you actually define the producer

dewan-ahmed replied:
With all the commits, the line# got all mixed up. This explanation is for the Avro part of the consumer/producer and is under the **Observation** section. The format is _I'm writing some code_ and then _here's what the code does_. 

Do you suggest I duplicate these texts to some places within the file? if yes, could you please comment the new line numbers? Thanks.

celestehorgan commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+- Create an AvroProducer instance.
+- Send a sample Avro message to the Kafka broker.
+
+Consumer program
```
Similar comment – put this content in the section you actually create the consumer in.

dewan-ahmed replied:
Same response https://github.com/aiven/devportal/pull/1733#discussion_r1136459492

celestehorgan reviewed (changes requested):
Lots of tiny grammar changes and a few bigger questions/changes for you :)

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+Observation
+"""""""""""
+
+You might have noticed ``key_deserializer``, ``key_serializer``, ``value_deserializer``, and ``value_serializer`` in these programs. Since Kafka brokers don't know about the records and only deal in bytes, the programs need to serialize 
+and deserialize data before making sense of them. 
```
Good callout. This is a residue from a previous version of the code where these variables were used. I took them out but didn't notice this line. I'm taking this line out now.

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+        end
+        p0 -->co0 
+
+If the first consumer ``co0`` crashes for some reason, the other consumer ``co1`` in the consumer group will begin consuming messages from the last committed offset of the partition. 
```
Did you mean "..cluster highly available."?

dewan-ahmed replied:
Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+            ssl_keyfile="service.key", # From the connection information for managed service
+        )
+
+        # Continuously poll for new messages
+        for msg in consumer:
+          print("Message: ", msg.value)
```
Good call. I added these lines under the **Observation** section. Let me know if you think there's a better place for this explanation.

dewan-ahmed replied:
Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389.

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+Overview
+--------
+
+Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
```
This is an insightful comment. Please check if the change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389 make sense.

dewan-ahmed commented on docs/tutorials/talk-to-kafka.rst:
```diff
@@ -0,0 +1,715 @@
+--------
+
+Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
+If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_, allowing you to start producing and consuming data in minutes.
```
Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389.

dewan-ahmed commented:
Many thanks @celestehorgan for detailed and insightful review. I've addressed most comments via accepting suggestions, https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389, and  https://github.com/aiven/devportal/pull/1733/commits/3a9a11af30ad0b1f64651cab3e34dcd079dba15c.

I'm not too sure about https://github.com/aiven/devportal/pull/1733#discussion_r1135265252 and the following question so I requested some clarification via https://github.com/aiven/devportal/pull/1733#discussion_r1136459492. I'll address this once I get the response. Thanks again.

dewan-ahmed commented:
Took a first pass at reST-->MD conversion via https://github.com/aiven/devportal/pull/1733/commits/532e037a941e2bca9458aaa35928d74513a237bc. Even with pandoc, there are significant manual work for the conversion. Next ToDo are:
- to fix all occurrences of `{.interpreted-text role="ref"}` and properly reference these sections
- to fix the formatting of the **Reference** section

celestehorgan commented on docs/tutorials/talk-to-kafka.md:
```diff
@@ -0,0 +1,786 @@
+
+## Overview
+
+Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. Have you added multiple Kafka brokers to ensure high availability while maintaining a low network
```
```suggestion
Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, standing up a production-ready Kafka cluster with security and high-availability is a different story. Have you added multiple Kafka brokers to ensure high availability while maintaining a low network
```

celestehorgan commented on docs/tutorials/talk-to-kafka.md:
```diff
@@ -0,0 +1,786 @@
+# Let's talk to Apache Kafka®: sending and receiving application data
+
+> **Note**
+> This tutorial assumes basic knowledge of Apache Kafka.
```
```suggestion
> This tutorial assumes basic knowledge of Apache Kafka®.
```

celestehorgan commented on docs/tutorials/talk-to-kafka.md:
```diff
@@ -0,0 +1,786 @@
+In this tutorial, we will learn how to create a highly available Apache
+Kafka service, go over the common tasks of producing and consuming
+messages, and finally use the popular Apache Avro™ specification to
+communicate with your Kafka service. This tutorial will use Python
```
```suggestion
communicate with your Kafka service. This tutorial will use the Python
```

TibsAtWork commented:
This is now at https://aiven.io/developer/how-to-send-and-receive-application-data-from-apache-kafka and so this PR can be closed
